numpy/__init__.pyi

```diff
@@ -2193,18 +2194,191 @@ class ndarray(_ArrayOrScalarCommon, Generic[_ShapeType, _DType_co]):
     ) -> Any: ...
 
     # `np.generic` does not support inplace operations
-    def __iadd__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __isub__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __imul__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __itruediv__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __ifloordiv__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __ipow__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __imod__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __ilshift__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __irshift__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __iand__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __ixor__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
-    def __ior__(self: _ArraySelf, other: ArrayLike) -> _ArraySelf: ...
+    @overload  # type: ignore[misc]
+    def __iadd__(self: _ArrayND[Any], other: _NestedSequence[str | bytes]) -> NoReturn: ...
+    @overload
+    def __iadd__(self: _ArrayND[bool_], other: _ArrayLikeBool_co) -> _ArrayND[bool_]: ...
+    @overload
+    def __iadd__(self: _ArrayND[unsignedinteger[_NBit1]], other: _ArrayLikeUInt_co) -> _ArrayND[unsignedinteger[_NBit1]]: ...
+    @overload
+    def __iadd__(self: _ArrayND[signedinteger[_NBit1]], other: _ArrayLikeInt_co) -> _ArrayND[signedinteger[_NBit1]]: ...
+    @overload
+    def __iadd__(self: _ArrayND[floating[_NBit1]], other: _ArrayLikeFloat_co) -> _ArrayND[floating[_NBit1]]: ...
+    @overload
+    def __iadd__(self: _ArrayND[complexfloating[_NBit1, _NBit1]], other: _ArrayLikeComplex_co) -> _ArrayND[complexfloating[_NBit1, _NBit1]]: ...
+    @overload
+    def __iadd__(self: _ArrayND[timedelta64], other: _ArrayLikeTD64_co) -> _ArrayND[timedelta64]: ...
+    @overload
+    def __iadd__(self: _ArrayND[datetime64], other: _ArrayLikeTD64_co) -> _ArrayND[datetime64]: ...
+    @overload
+    def __iadd__(self: _ArrayND[object_], other: Any) -> _ArrayND[object_]: ...
+    @overload
+    def __iadd__(self: _ArrayND[_ScalarType], other: _RecursiveSequence) -> _ArrayND[_ScalarType]: ...
+
+    @overload  # type: ignore[misc]
+    def __isub__(self: _ArrayND[Any], other: _NestedSequence[str | bytes]) -> NoReturn: ...
+    @overload
+    def __isub__(self: _ArrayND[bool_], other: _ArrayLikeBool_co) -> NoReturn: ...
```

Comment:
This seems odd, `__iadd__` and `__imul__` return `_ArrayND[bool_]` for this combination of input types.

Reply:
This because the subtraction of two boolean arrays raises an error, hence the [`NoReturn`](https://docs.python.org/3/library/typing.html#typing.NoReturn).
Come to think of it, the `NoReturn` is actually redundant here; removing the overload accomplishes the same.

Reply:
Removed as of https://github.com/numpy/numpy/pull/18417/commits/c6bdbad5458c3b48ee92350fa0b723ee3b98743f.